models/User.js

```diff
@@ -0,0 +1,54 @@
+const mongoose = require('mongoose')
+const Schema = mongoose.Schema
+const bcrypt = require('bcryptjs')
+const jwt = require('jsonwebtoken')
+
+
+const UserSchema = new Schema({  name: {
+    type: String,
+    required: [true, 'Please provide name'],
+    maxlength: 50,
+    minlength: 3,
+  },
+  email: {
+    type: String,
+    required: [true, 'Please provide email'],
+    match: [
+      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
+      'Please provide a valid email',
+    ],
+    unique: true,
+  },
+  password: {
+    type: String,
+    required: [true, 'Please provide password'],
+    minlength: 8,
+  },
+})
+
+UserSchema.pre('save', async function(next) {
+    const user = this;
+
+// only hash the password if it has been modified (or is new)
+if (!user.isModified('password')) return next();
+
+// generate a salt
+const salt = await bcrypt.genSalt(10);
+this.password = await bcrypt.hash(this.password, salt)
+next()
+
+});
+
+
+UserSchema.methods.createJWT = function () {
+  return jwt.sign({userId: this._id, name: this.name }, process.env.JWT_SECRET, {
+    expiresIn: process.env.JWT_LIFETIME,
+  })
+}
+
+UserSchema.methods.comparePassword = async function (canditatePassword) {
+  const isMatch = await bcrypt.compare(canditatePassword, this.password)
+  return isMatch
+}
```

Comment:
You asked if `await` was really necessary here. Honestly, `Promises` are _much_ easier to reason about when using **TypeScript** because you more verbosely define the method _contract_ (i.e. what parameters the method expects, and what value the method returns). 😅 

So when in pure JS land, you have to more proactively think explicitly about those things on your own. Why does VSCode say the `await` is unnecessary, and why did it work?

`comparePassword` is an `async` function. What does that really mean? It is going to return a `Promise`, and in this case, it is going to return a `Promise<boolean>`... a promise that resolves to a boolean if the passwords matched. Because there is no additional handling after you `await` the return value of `bcrypt.compare`, the value gets re-wrapped in a `Promise` and delivered to wherever it was invoked, currently in the `login` method where you `await` it to resolve the value again! 

So because you (1) don't do anything with the resolved promise and (2) return it to be resolved again later on, it is indeed not necessary here.

You could even go one step further - the function itself doesn't have to be async since `bcrypt.compare` returns a Promise already.
```javascript
UserSchema.methods.comparePassword = function (canditatePassword) { 
   return bcrypt.compare(canditatePassword, this.password);
}
```